perf-tool/src/server.cpp

```diff
@@ -0,0 +1,310 @@
+            {
+                command = jsonCommand.dump();
+                handleClientCommand(command, "Commands file");
+            }
+        }
+
+        /* Receiving messages to clients */
+        for (int i = 0; i < activeNetworkClients; i++)
+        {
+            command = networkClients[i]->handlePoll();
+            if (!command.empty())
+            {
+                handleClientCommand(command, "Client");
+            }
+        }
+
+        /* Checks if it is the time for any delayed command to be fired */
+        if (!delayedCommands.empty()) {
+            auto currentClockTime = std::chrono::system_clock::now();
+            std::time_t currentTime = std::chrono::system_clock::to_time_t(currentClockTime);
+
+            while(delayedCommands.size() > 0)
+            {
+                if (delayedCommands[0].delayTill <= currentTime) {
+                    agentCommand(delayedCommands[0].command);
+                    delayedCommands.erase(delayedCommands.begin());
+                } else{
+                    break;
+                }
+            }
+        }
+    }
+}
+
+void Server::execCommand(json command)
+{
+    if (command.find("delay") != command.end() && command["delay"].get<int>() > 0) {
+        auto currentClockTime = std::chrono::system_clock::now();
+        std::time_t currentTime = std::chrono::system_clock::to_time_t(currentClockTime);
+        std::time_t delay_till = currentTime + command["delay"].get<int>();
+
+        delayed_command_t delayedCommand;
+        delayedCommand.command = command;
+        delayedCommand.delayTill = delay_till;
+        delayedCommands.push_back(delayedCommand);
+
+        sort(delayedCommands.begin(), delayedCommands.end(),
+            [](delayed_command_t a, delayed_command_t b) {
+                if (a.delayTill < b.delayTill) {
+                    return true;
+                } else {
+                    return false;
+                }
+        });
+    }
+    else if ((command["functionality"].get<std::string>()).compare("perf"))
+    {
+        agentCommand(command);
+    }
+    else
+    {
+        startPerfThread(command["time"]);
+    }
+}
+
+void Server::handleClientCommand(string command, string from)
+{
+    json com;
+    try
+    {
+        com = json::parse(command);
+        execCommand(com);
+    }
+    catch (const std::exception &e)
+    {
+        std::cerr << e.what() << '\n';
+        std::cerr << "Improper command received from: " << from << '\n';
+    }
+
+    loggingClient->logData(command, from);
+}
+
+void Server::sendMessage(const int socketFd, const string message)
+{
+    int n, total = 0;
+    size_t length = message.size();
+    const char *buffer = message.data();
+
+    while (total < length) 
+    {
+        n = send(socketFd, buffer, strlen(buffer), 0);
+        if (n == -1) 
+        {
+            perror("ERROR sending message to clients failed"); 
+        }
+
+        total += n;
+        buffer += n;
+    }
```

Comment:
If the client goes away the server will continue to send data over and over again resulting in countless messages like:
`ERROR sending message to clients failed: Broken pipe`